clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1053,6 +1048,14 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
         return new ClusterAndWaitTime(cluster, elapsed);
     }
 
+    private TimeoutException timeoutForAbsentMetadata(String topic, Integer partition, long maxWaitMs, Integer partitionsCount) {
+        return new TimeoutException(partitionsCount == null ?
+                String.format("Topic %s not present in metadata after %d ms.",
+                        topic, maxWaitMs) :
+                String.format("Partition %d of topic %s with partition count %d is not present in metadata after %d ms.",
+                        partition, topic, partitionsCount, maxWaitMs));
```

Comment:
Should we set a cause to the `TimeoutException` to make it easy to programmatically handle the case where the topic or partition doesn't exist (vs all other reasons why `TimeoutException` may be thrown).

Reply:
+1 this comment.  It would be really useful to have a cause that indicates the reason for this TimeoutException, without having to check the error message.